spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/EntityPropertyValueProvider.java

```diff
@@ -51,14 +65,37 @@
 	@SuppressWarnings("unchecked")
 	private <T> T readSingleWithConversion(DatastorePersistentProperty persistentProperty) {
 		Object val = this.entity.getValue(persistentProperty.getFieldName()).get();
-		Class<?> targetType = persistentProperty.getType();
 		if (val == null) {
 			return null;
 		}
-		if (ClassUtils.isAssignable(targetType, val.getClass())) {
-			return (T) val;
+
+		Class<?> targetType = persistentProperty.getType();
+		T result = null;
+
+		DefaultDatastoreEntityConverter.TwoStepConversion twoStepConversion = this.conversionGetter.apply(targetType);
```

Comment:
Why don't we have a method/interface that simply takes an `Object` and a `targetType` and does the conversion?

Reply:
mostly for symmetry with value write logic, but we can put all this in one class